drivers/hmis_external_apis/app/jobs/hmis_external_apis/ac_hmis/create_referral_job.rb

```diff
@@ -165,35 +163,54 @@ def update_client_addresses(client)
           ),
           AddressID: Hmis::Hud::Base.generate_uuid,
           **common_client_attrs(client),
-        }
+        )
+      end
+
+      # See if any values are equal to existing values. If so, update the existing value.
+      old_addresses = client.addresses
+      new_addreseses.each do |addy|
+        addy.id = old_addresses.find { |old| old.equal_for_merge?(addy) }&.id
       end
-      Hmis::Hud::CustomClientAddress.import!(client_address_attrs)
+
+      Hmis::Hud::CustomClientAddress.import!(
+        new_addreseses,
+        on_duplicate_key_update: { conflict_target: [:id], columns: [:DateUpdated] },
+      )
     end
 
+    # Add/update phones and emails
     def update_client_contacts(client)
-      # replace old phones, and emails
-      client.contact_points.destroy_all
-
-      client_phone_attrs = params[:phone_numbers].to_a.map do |values|
-        {
+      new_contact_points = []
+      params[:phone_numbers].to_a.each do |values|
+        new_contact_points << Hmis::Hud::CustomClientContactPoint.new(
           system: :phone,
           value: values[:number],
-          **values.slice(:use, :notes),
+          use: values[:type],
+          notes: values[:notes],
           ContactPointID: Hmis::Hud::Base.generate_uuid,
           **common_client_attrs(client),
-        }
+        )
       end
-      Hmis::Hud::CustomClientContactPoint.import!(client_phone_attrs)
 
-      client_email_attrs = params[:email_address].to_a.map do |value|
-        {
+      params[:email_address].to_a.each do |value|
+        new_contact_points << Hmis::Hud::CustomClientContactPoint.new(
           system: :email,
           value: value,
           ContactPointID: Hmis::Hud::Base.generate_uuid,
           **common_client_attrs(client),
-        }
+        )
+      end
+
+      # See if any values are equal to existing values. If so, update the existing value.
+      old_contact_points = client.contact_points
+      new_contact_points.each do |cp|
+        cp.id = old_contact_points.find { |old| old.equal_for_merge?(cp) }&.id
       end
-      Hmis::Hud::CustomClientContactPoint.import!(client_email_attrs)
+
+      Hmis::Hud::CustomClientContactPoint.import!(
+        new_contact_points,
+        on_duplicate_key_update: { conflict_target: [:id], columns: [:notes, :DateUpdated] },
```

Comment:
If the element has a note but we don't receive one, this nullifies the old note. That's not great. Could set `cp.note ||= found_match.note` above. I don't think there's a way to specify it for on_duplicate_key_update

Reply:
So, you could do that up on line 207, and only merge the note field if it's not null, BUT that will prevent ever deleting a note...

Reply:
I updated the logic to be much simpler, let me know what you think. If we get a phone # from LINK that we already have stored, we will update the record. However if there is already a note on that phone number, and LINK didn't send one, we will leave the old note. I think that is the desired behavior.. we don't want LINK overwriting notes that users have made in the HMIS. AND we don't want a ton of duplicate values. 


> that will prevent ever deleting a note...
The user can always delete a note in the UI, I think that is sufficient?